Implement Web scraping "HTML + XPath"

Alkarex opened the pull request:
This PR adds a (killer? 🤩) functionality, namely the ability to consume any Web site / HTML source, also when an RSS / Atom feed is not available.

It is based on XPath 1.0, which is natively available in PHP:
* https://www.w3.org/TR/xpath-10/
* https://php.net/class.domxpath
* https://developer.mozilla.org/docs/Web/XPath/Axes

This is a light version of what RSS Bridge offers, but natively inside FreshRSS and much easier to use:
* https://github.com/RSS-Bridge/rss-bridge/wiki/The-collectData-function
* https://github.com/RSS-Bridge/rss-bridge/wiki/XPathAbstract

Using a third-party tool did not scale enough anymore for me, when I just need to quickly add another (simple) source.

FreshRSS can then also republish as RSS.

This can be combined with our existing ability to follow article links to retrieve full article content (use with parsimony).

Contributes indirectly to other issues:
* Better handling of enclosures
* Better RSS outputs for https://github.com/FreshRSS/FreshRSS/issues/1796
* Cache HTTP requests when getting full article content
* Better handling of encoding when getting full article content
* Purge old feed caches regularly
* Keep the original order of articles when they all have the same date (before, we got a random order)
* More PHP type hints
* Typos in French

What remains to be done (other PRs):
* [ ] Preview to help writing the XPath expressions just like for our full article content retrieval (help welcome as I will not be able to do much in the coming two weeks)
* [ ] Consume "JSON +XPath" sources (only very little needs to be changed to support that)
* [x] A few new TODOs added in the code (in particular an easy performance improvement related to iterating over SimplePie entries) https://github.com/FreshRSS/FreshRSS/pull/4263
* [ ] Documentation (help welcome)
  * Technical note: all XPath evaluations (i.e. XPath supposed to return some text and not the queries returning a collection) are automatically wrapped into a [`normalize-space()`](https://developer.mozilla.org/docs/Web/XPath/Functions/normalize-space) function to cast the result to string if not already done by the user and also remove superfluous white spaces.
* [x] Include information in OPML export / import? https://github.com/FreshRSS/FreshRSS/pull/4342
* [ ] Custom date format, e.g. with https://php.net/datetime.createfromformat
* [x] Make corresponding upstream PR, reverting https://github.com/simplepie/simplepie/commit/e49c578817aa504d8d05cd7f33857aeda9d41908#commitcomment-67585150 : https://github.com/simplepie/simplepie/pull/723

Summary of the tasks from comments below:
* [x] Implement support of `<base>`

Alkarex committed:
More PHP type hints for Fever
Follow-up of https://github.com/FreshRSS/FreshRSS/pull/4201
Related to https://github.com/FreshRSS/FreshRSS/issues/4200

Alkarex committed:
Detail

Alkarex committed:
Merge branch 'edge' into source-scrape-xpath

Alkarex committed:
Draft

Alkarex committed:
Merge branch 'edge' into source-scrape-xpath

Alkarex committed:
Merge branch 'edge' into source-scrape-xpath

Alkarex committed:
Progress

Alkarex committed:
More draft

Alkarex committed:
Fix thumbnail PHP type hint
https://github.com/FreshRSS/FreshRSS/issues/4215

Alkarex committed:
Merge branch 'edge' into source-scrape-xpath

Alkarex committed:
Merge branch 'edge' into source-scrape-xpath

Alkarex committed:
More types

Alkarex committed:
A bit more

Alkarex committed:
Refactor FreshRSS_Entry::fromArray

Alkarex committed:
Progress

Alkarex committed:
Starts to work

Alkarex committed:
Categories

Alkarex committed:
Fonctional

Alkarex committed:
Layout update

Alkarex committed:
Fix relative URLs

Alkarex committed:
Cache system

Alkarex committed:
Forgotten files

Alkarex committed:
Merge branch 'edge' into source-scrape-xpath

Alkarex commented:
Example for `https://www.france.tv/france-2/les-petits-meurtres-d-agatha-christie/toutes-les-videos/`

![image](https://user-images.githubusercontent.com/1008324/153773031-68ddf8e9-7c29-4f1c-bd91-f30c84c584a1.png)

![image](https://user-images.githubusercontent.com/1008324/153772998-541c1d4b-1865-4631-9027-9f289b04c709.png)

Result:
![image](https://user-images.githubusercontent.com/1008324/153772891-13d6a75f-2936-46d1-968b-f6e138518f84.png)

Alkarex commented on app/Models/Feed.php:
```diff
@@ -489,6 +542,87 @@ public function loadEntries(SimplePie $simplePie) {
+	 * @param array<string,mixed> $attributes
+	 * @return SimplePie|null
+	 */
+	public function loadHtmlXpath(bool $loadDetails = false, bool $noCache = false, array $attributes = []) {
```
This function is the only true new functional code for this feature, the rest of the PR being mainly related improvements and UI code

Alkarex commented on app/views/index/rss.phtml:
```diff
@@ -27,12 +36,23 @@ foreach ($this->entries as $item) {
 				}
+				$enclosures = $item->enclosures(false);
+				if (is_array($enclosures)) {
+					foreach ($enclosures as $enclosure) {
+						// https://www.rssboard.org/media-rss
+						echo "\t\t\t" , '<media:content url="' . $enclosure['url']
+							. (empty($enclosure['medium']) ? '' : '" medium="' . $enclosure['medium'])
+							. (empty($enclosure['type']) ? '' : '" type="' . $enclosure['type'])
+							. (empty($enclosure['length']) ? '' : '" length="' . $enclosure['length'])
+							. '"></media:content>', "\n";
```
https://github.com/FreshRSS/FreshRSS/issues/1796

Alkarex commented on lib/lib_rss.php:
```diff
@@ -218,6 +218,7 @@ function customSimplePie($attributes = array()): SimplePie {
 	$simplePie->set_cache_name_function('sha1');
 	$simplePie->set_cache_location(CACHE_PATH);
 	$simplePie->set_cache_duration($limits['cache_duration']);
+	$simplePie->enable_order_by_date(false);
```
We sort ourselves. This created a bug randomising the order of entries when several have the same date.
Also improves performances slightly by avoiding a sort.

Alkarex committed:
Remove a debug line

Alkarex commented:
Another example. Test page: https://alexandre.alapetite.fr/blog/actu.fr.html

Reading from a basic HTML structure
```html
<dl>
  <dt id="entry1">Entry 1 title</dt>
  <dd>Entry 1 content</dd>
  <dt id="entry2">Entry 2 title</dt>
  <dd>Entry 2 content</dd>
</dl>
```

![image](https://user-images.githubusercontent.com/1008324/153777098-54204f54-2a3a-4d22-afe2-86e72f1fd502.png)

Frenzie commented on app/Models/Feed.php:
```diff
@@ -428,15 +478,18 @@ public function loadEntries(SimplePie $simplePie) {
+							$enclosureContent .= '<p class="enclosure-content"><a download="" href="' . $elink
+								. ($mime == '' ? '' : '" data-type="' . htmlspecialchars($mime, ENT_COMPAT, 'UTF-8'))
+								. ($medium == '' ? '' : '" data-medium="' . htmlspecialchars($medium, ENT_COMPAT, 'UTF-8'))
+								. '">💾</a></p>';
```
Not related to the PR but this caught my eye due to being in the diff. It'd probably make sense to stick in an aria-label="Save" or some such. Cf. https://github.com/FreshRSS/FreshRSS/issues/4190

Alkarex replied:
There is already the `download` attribute, which can be used by assistive tools to know the intent. Maybe a `title` though, to help understanding the floppy disk icon for both screen and screen-reader users...

Alkarex committed:
Automatic form validation of XPath expressions

Alkarex commented:
Now with automatic form validation of XPath expressions
![image](https://user-images.githubusercontent.com/1008324/153913526-976a3b20-46f9-43ab-befd-eaf79bc5dffc.png)

Alkarex committed:
data-leave-validation

Alkarex committed:
Fix reload action

Alkarex committed:
Simpler examples

Alkarex committed:
Fix column type for PostgreSQL

CE1212 commented:
This is something I have been looking for...Great

Alkarex committed:
Enforce HTTP encoding

Alkarex commented:
> This is something I have been looking for...Great

@CE1212 Tests welcome ;-)

Alkarex committed:
Readme

CE1212 commented:
So would it be possible to have it get the RSS then go to the site and scrape further information, say we only get the headline and a link, so the Freshrss once ready could go to the page and get say the author link also once its on the site.

If it could do that and rotate proxies that would be amazing.

Can't wait for it to be ready...awesome stuff

Alkarex commented:
@CE1212 This feature is an alternative to an RSS / Atom feed, not a supplement. To supplement an existing RSS / Atom feed, FreshRSS already has a full-content retrieval mechanism. I am not sure what you mean with proxy rotation, but it is likely not related.

CE1212 commented:
@Alkarex 
Ah I just totally misunderstood, the proxy rotation I meant having the script cycle through a series of ip's in case you are accessing the same site really often.

I don't seem to have the type of feed source option on my install

I'm running
Your current version of FreshRSS is 1.20.0-dev.

unless I am just not seeing it probably me doing something wrong but any help would be appreciated 

I am happy to test as I know enough just to break things lol, thanks for your replies

Edit-
I had the wrong version installed, I do only know enough to break things.....thanks for your hard work

Alkarex committed:
Fix get full content

math-GH commented:
I play a little bit around with this PR.

Here some findings:
found this warning in my log
![grafik](https://user-images.githubusercontent.com/1645099/155383002-daa48d33-18a9-4862-a2a4-bd2961e95058.png)

Git does not ignore the cached HTML file in data/cache
![grafik](https://user-images.githubusercontent.com/1645099/155383271-ec3cf310-4985-41cc-8267-67578c46a11e.png)


What does the orange mean?
![grafik](https://user-images.githubusercontent.com/1645099/155383481-d9ea4a81-4005-4b74-99fc-e06b2ebed34b.png)
![grafik](https://user-images.githubusercontent.com/1645099/155383533-e39ac8ec-3805-47c3-9f26-2c4dd4902f52.png)

math-GH commented:
Could you please try, whether you are able to generate a feed from this page?

https://contao.org/en/news.html?year=2021

my settings:
![grafik](https://user-images.githubusercontent.com/1645099/155387705-098ec7ca-8dfd-469e-9a30-77a2667af326.png)

I am not sure if XPath has issues with HTML5 and its not closed tags like f.e. `<meta` or `<link>`.
I tried XPath on https://www.freeformatter.com/xpath-tester.html and it gives error messages because of missed `</meta>` and `</link>`

math-GH commented:
the XPath for feed title is used only once: when the feed is added new to the system.

When the feed config is changed it does not have any affect on the title anymore
![grafik](https://user-images.githubusercontent.com/1645099/155393213-81165a97-fc95-4953-a5d8-5bd5a7e74555.png)

Frenzie commented:
For me that'd very much be the desired behavior; I wouldn't want it randomly changing the name into something I didn't put in there.

math-GH commented:
> For me that'd very much be the desired behavior; I wouldn't want it randomly changing the name into something I didn't put in there.

So it would make sense to hide the config textarea if it has no affect anymore.

Frenzie commented:
> I am not sure if XPath has issues with HTML5 and its not closed tags like f.e. <meta or <link>.

The data should be fed through https://www.php.net/manual/en/domdocument.loadhtml.php

For a tool like you linked you'd have to do that yourself, i.e., change `<meta>` into a self-closing `<meta/>` for example.

Frenzie commented:
Anyway, here's why `querySelectorAll()`'s sometimes nicer than XPath. I figure something like this is what we want.
```
//div[contains(concat(' ', @class, ' '), 'odd') or contains(concat(' ', @class, ' '), 'even')]
```
[Edit: sorry, overlooked that I'd need to stick in even more spaces, something like shown in <https://stackoverflow.com/a/1390680>]

![Screenshot_2022-02-23_22-41-24](https://user-images.githubusercontent.com/202757/155413374-8281dd66-a70e-41dd-9da2-d640377c1461.png)

![Screenshot_2022-02-23_22-40-57](https://user-images.githubusercontent.com/202757/155413390-67c2ad3d-1d5a-4f52-8f1b-a01190665ee9.png)

math-GH commented:
The link to strtotime() could be more useful with a `target="_blank"` so that the input is not lost while editing and loading a new page in the same tab.
![grafik](https://user-images.githubusercontent.com/1645099/155606163-0d4f7573-ceab-47a5-902f-7ed7bec52f72.png)

math-GH commented:
strtotime() is only useful for English pages. Maybe there could be a way to have customized dates

Maybe https://www.php.net/manual/en/datetime.createfromformat.php could be an idea

math-GH commented:
![grafik](https://user-images.githubusercontent.com/1645099/155608326-23f7c076-5d0c-402f-ad0d-628117389b00.png)
It "eats" `&amp;cHash`

HTML on the page:
![grafik](https://user-images.githubusercontent.com/1645099/155608572-d2ea2c88-6cb1-4910-b38a-8633028d88f1.png)

but the article link is (see right hand side of `5831`:
`....../?tx_ttnews%5Btt_news%5D=5831=863ee361abe89b13ebcef16ee0b3024a`

math-GH commented:
The creation of links works not best.

"Feed Url": `https://domain.tld/sub/`
but: `<base href="https://domain.tld" />`
links on the page: `/sub/newsarticle-1234.html`

The generated link: `https://domain.tld/sub/sub/newsarticle-1234.html`.
Expected link: `https://domain.tld/sub/newsarticle-1234.html`

This functionality needs to search and follow the `<base>`

Frenzie commented:
Is that from the source or from the browser? That is, the XPath output would have performed an `&amp;` → `&` transformation, but logically then you'd still have `&cHash` remaining.

Edit: of course there would heed to be an htmlspecialchars() applied to prevent anything weird from happening in the browser, not talking about that part. ^_^

Frenzie commented on app/Models/Feed.php:
```diff
@@ -489,6 +542,95 @@ public function loadEntries(SimplePie $simplePie) {
+		if ($this->url == '') {
+			return null;
+		}
+		$feedSourceUrl = htmlspecialchars_decode($this->url, ENT_QUOTES);
+		if ($this->httpAuth != '') {
+			$feedSourceUrl = preg_replace('#((.+)://)(.+)#', '${1}' . $this->httpAuth . '@${3}', $feedSourceUrl);
+		}
```
Cf. discussion in https://github.com/FreshRSS/FreshRSS/pull/4220#issuecomment-1050289052

I don't have time to check right now but there might be something double going on? I think the XPath output is already something akin to `htmlspecialchars_decode()` out of the box, unless I'm mistaken.

Frenzie replied:
@math-GH Btw, what's a sample page? I don't really want to create a testcase myself. ;-)

math-GH replied:
see my next comment

Alkarex replied:
https://github.com/FreshRSS/FreshRSS/pull/4220/commits/998babbe94755c832a226c79527abbd5b466bbe3
The rest looks good

math-GH commented:
From source: `wasserball/aktuelles-wasserball/lesen/?tx_ttnews%5Btt_news%5D=5831&cHash=863ee361abe89b13ebcef16ee0b3024a`

math-GH commented:
I do not want to publish and archive concrete links, so I would use an etherpad to share the link and configs:
https://yopad.eu/p/xpath-365days

math-GH commented:
If someone needs an XPath sandbox to find the correct XPath rules: http://xpather.com/ (maybe we could add it later into the documentation)

Frenzie commented:
Cheers, that would've been nice to have 20 years ago. (It seems to date back to 2018; I actually haven't used XPath/XSLT since 2015 iirc.)

Alkarex committed:
Merge branch 'edge' into source-scrape-xpath

Alkarex committed:
target="_blank"

Alkarex committed:
gitignore

Alkarex commented:
I am not fully back yet, so just a few quick things:

> found this warning in my log

Normal. A new SQL column is automatically added on demand.

> Git does not ignore the cached HTML file in data/cache

Done: https://github.com/FreshRSS/FreshRSS/pull/4220/commits/839270835b2457fc1f705d7b957f0062dcf2b283

> What does the orange mean?

Orange is not related to this PR but is probably when your feed is empty.

> The link to strtotime() could be more useful with a `target="_blank"` so that the input is not lost while editing and loading a new page in the same tab.

Done: https://github.com/FreshRSS/FreshRSS/pull/4220/commits/aa7901f4e119389b705b428607811249f3119774

Alkarex commented:
> strtotime() is only useful for English pages. Maybe there could be a way to have customized dates
> 
> Maybe https://www.php.net/manual/en/datetime.createfromformat.php could be an idea

Good idea with custom format, but let's keep that for another PR. `strtotime()` is not only for English dates, but also for e.g. ISO and other formats using only numbers

Frenzie commented:
As an aside, many pages use a proper `<time>` attribute, so you should use the `datetime` attribute while ignoring the text content. E.g.,

```
<time datetime="2022-02-19T11:34:23+01:00">February 19, 2022 at 11:34</time>
```

Alkarex commented:
> Could you please try, whether you are able to generate a feed from this page? https://contao.org/en/news.html?year=2021

A simpler version based on @Frenzie 's version:

![image](https://user-images.githubusercontent.com/1008324/155878763-3acd828b-e23f-4df9-a331-670ee623a32e.png)

Alkarex commented:
> Anyway, here's why `querySelectorAll()`'s sometimes nicer than XPath.

@Frenzie , yes, CSS selectors are easier (but less expressive) than XPath. I believe we should have both. I started with XPath because there is native PHP support (of XPath 1.0) and because I believe our future "HTML + CSS Selector" scraping method could behind the scenes perform a CSS Selector to XPath translation (I believe there are some light libraries doing that - findings welcome). That could also be used to remove our `lib_phpQuery.php`

Alkarex committed:
htmlspecialchars_utf8

Alkarex commented:
> It "eats" `&amp;cHash`

@math-GH Indeed. Should be fixed by https://github.com/FreshRSS/FreshRSS/pull/4220/commits/998babbe94755c832a226c79527abbd5b466bbe3

Alkarex committed:
Implement HTML <base>
And fix/revert `xml:base` support in SimplePie https://github.com/simplepie/simplepie/commit/e49c578817aa504d8d05cd7f33857aeda9d41908

Alkarex commented:
> This functionality needs to search and follow the `<base>`

@math-GH Should be fixed by https://github.com/FreshRSS/FreshRSS/pull/4220/commits/f3f92aa1889950d543e4c0ac1ecd5e4c59aa37df . Tests welcome

Related upstream PR: https://github.com/simplepie/simplepie/pull/723

Alkarex committed:
Merge branch 'edge' into source-scrape-xpath

math-GH commented:
I love it. It works very fine now.

Alkarex committed:
SimplePie upstream PR merged
https://github.com/simplepie/simplepie/pull/723

Alkarex commented:
As all seems fine, let's merge and test a bit more

Alkarex commented on lib/lib_phpQuery.php:
```diff
@@ -436,7 +436,8 @@ protected function isXHTML($markup = null) {
 //		return strpos($markup, '<?xml') !== false && stripos($markup, 'xhtml') === false;
-		return strpos(substr($markup, 0, 100), '<'.'?xml') !== false;
+		$head = substr($markup, 0, 100);
+		return strpos($head, '<'.'?xml') !== false && stripos($head, '<html ') === false;
```
https://github.com/FreshRSS/FreshRSS/pull/4243

Alkarex commented:
> @Frenzie , yes, CSS selectors are easier (but less expressive) than XPath. I believe we should have both. I started with XPath because there is native PHP support (of XPath 1.0) and because I believe our future "HTML + CSS Selector" scraping method could behind the scenes perform a CSS Selector to XPath translation (I believe there are some light libraries doing that - findings welcome). That could also be used to remove our `lib_phpQuery.php`

https://github.com/FreshRSS/FreshRSS/pull/4261

InterferencePattern commented:
When will this feature be included in a release?

Alkarex commented:
> When will this feature be included in a release?

@jimbudarz You are welcome to already use our edge branch, which is a relatively stable rolling release, available by git, Docker, ZIP. Feedback welcome, and we need more testers. Next stable version 1.20.0 is coming in I guess ~2-3 weeks.

math-GH commented:
I can confirm that the edge (V1.20.0-dev) works very well. I use it on my prod instance. The Web scraping works great there.

Alkarex commented:
Nice article on the subject: https://danq.me/2022/09/27/freshrss-xpath/

_Originally posted by @marienfressinaud in https://github.com/FreshRSS/FreshRSS/discussions/4647#discussioncomment-3777171_

math-GH commented:
More amazing feedback: https://forum.cloudron.io/topic/7651/freshrss-1-2-0-released-with-killer-feature-track-any-website

Alkarex commented on app/Models/Feed.php:
```diff
@@ -489,6 +542,97 @@ public function loadEntries(SimplePie $simplePie) {
+			return null;
+		}
+
+		$html = getHtml($feedSourceUrl, $attributes);
```
Bug fix `$attributes` fixed in https://github.com/FreshRSS/FreshRSS/pull/4759
